bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs/diff/VcsServer.scala

```diff
@@ -0,0 +1,171 @@
+
+  def lookup(id: SPProgramID): VcsAction[ISPProgram] =
+    (Option(odb.lookupProgramByID(id)) \/> NotFound(id)).liftVcs
+
+
+  /** Creates an action that reads from a program with a read lock held
+    * provided the caller has permission. */
+  def read[A](id: SPProgramID, user: Set[Principal])(body: ISPProgram => A): VcsAction[A] =
+    managed(id, user, readLock, readUnlock)(p => body(p).point[VcsAction])
+
+  /** Creates an action that potentially writes to a program with a write lock
+    * held provided the caller has permission.  Writing is done on a copy of the
+    * associated program and only when successful the copy replaces the current
+    * version of the program in the database.
+    *
+    * The caller provides three functions, `evaluate`, `filter`, and `update`.
+    * Evaluate should not modify anything in the progrm but rather calculate
+    * a value.  The value is handed to `filter` to decide whether to continue
+    * with the write operation (and make the program copy).  In either case,
+    * the value is returned as the result of the action.  The `update` takes a
+    * copy of the program and the value returned by `evaluate` and mutates the
+    * copy.  If successful, the copy replaces the current version of the program
+    * in the database.
+    *
+    * @param id program id
+    * @param user authentication
+    * @param evaluate a function to compute a value which is used to determine
+    *                 whether to continue the write operation
+    * @param filter a function that takes the value returned by the `evaluate`
+    *               function and returns `true` if the write should continue
+    * @param update a function that performs the side effect of updating the
+    *               program copy
+    * @return the value produced by the `evaluate` function
+    */
+  def write[A](id:       SPProgramID,
+               user:     Set[Principal],
+               evaluate: ISPProgram => VcsAction[A],
+               filter:   A => Boolean,
+               update:   (ISPFactory, ISPProgram, A) => VcsAction[Unit]): VcsAction[A] =
+    managed(id, user, writeLock, writeUnlock) { prog =>
+      evaluate(prog) >>= { a =>
+        if (filter(a)) {
+          val cp = odb.getFactory.copyWithSameKeys(prog)
+          update(odb.getFactory, cp, a) >> putProg(cp).map(_ => a).liftVcs
+        } else {
+          VcsAction(a)
+        }
+      }
+    }
+
+  /** Adds the given program to the database, provided it doesn't share the
+    * same key or id with an existing program in the database and the user has
+    * appropriate keys.
+    */
+  def add(p: ISPProgram, user: Set[Principal]): VcsAction[Unit] = {
+    def failIfExists(id: SPProgramID, key: SPNodeKey): VcsAction[Unit] = {
+      val lookupFailure = Option(odb.lookupProgramByID(id)).map(_ => IdAlreadyExists(id)).orElse {
+        Option(odb.lookupProgram(key)).map(_ => KeyAlreadyExists(id, key))
+      }
+
+      (lookupFailure <\/ (())).liftVcs
```

Comment:
Ah, right. That seems fine. If you need to do it more than once you might add a `VcsAction.void` or `.unit` constant for the no-op action.